### ISSUE TITLE
High CPU load on `calico-node -felix`

### ISSUE THREAD

**Issue by sfudeus** [projectcalico/calico#4129]
## Expected Behavior

In a K8s cluster of <20 nodes, calico-nodes have a rather stable CPU usage at around 60m CPU and less than 250MiB RAM. 
In one of our larger clusters with 250 nodes, calico-nodes have stable CPU usage at usually < 300m CPU and less than 300MiB RAM. 

## Current Behavior

Since upgrading to calico 3.15 (maybe already 3.14) and using route reflectors, we observe that after our weekly rolling node redeploy process, most calico-node processes are reporting high CPU load (running with CPU limit 1). 

```
root      322520  0.0  0.0 110144 12948 ?        Sl   Oct27   0:44  \_ containerd-shim -namespace moby -workdir /var/lib/containerd/io.containerd.runtime.v1.linux/moby/d275fe7b1f97da40a0faa58b7c86f1286b78aeaec4721b34890a3ce974f6906b -address /run/docker/libc
root      322539  0.0  0.0   4364   800 ?        Ss   Oct27   0:00  |   \_ /usr/local/bin/runsvdir -P /etc/service/enabled
root      322659  0.0  0.0   4212   724 ?        Ss   Oct27   0:00  |       \_ runsv felix
root      322666 94.3  0.0 4758960 101344 ?      Rl   Oct27 952:53  |       |   \_ calico-node -felix
root      322660  0.0  0.0   4212   788 ?        Ss   Oct27   0:00  |       \_ runsv bird6
root      322961  0.1  0.0    760   604 ?        S    Oct27   1:25  |       |   \_ bird6 -R -s /var/run/calico/bird6.ctl -d -c /etc/calico/confd/config/bird6.cfg
root      322661  0.0  0.0   4212   728 ?        Ss   Oct27   0:00  |       \_ runsv monitor-addresses
root      322665  0.0  0.0 3796348 61344 ?       Sl   Oct27   0:14  |       |   \_ calico-node -monitor-addresses
root      322662  0.0  0.0   4212   724 ?        Ss   Oct27   0:00  |       \_ runsv allocate-tunnel-addrs
root      322668  0.1  0.0 3869568 69536 ?       Sl   Oct27   1:11  |       |   \_ calico-node -allocate-tunnel-addrs
root      322663  0.0  0.0   4212   852 ?        Ss   Oct27   0:00  |       \_ runsv bird
root      322962  0.1  0.0    936   760 ?        S    Oct27   1:42  |       |   \_ bird -R -s /var/run/calico/bird.ctl -d -c /etc/calico/confd/config/bird.cfg
root      322664  0.0  0.0   4212   736 ?        Ss   Oct27   0:00  |       \_ runsv confd
root      322670  0.2  0.0 4018568 80052 ?       Sl   Oct27   2:26  |           \_ calico-node -confd
```

Logs are inconspicuous, just a periodic (1m) note about using an autodetected IP address:

```
2020-10-28 09:28:09.801 [INFO][68] monitor-addresses/startup.go 756: Using autodetected IPv4 address on interface eno1: 10.238.1.4/24
2020-10-28 09:29:09.888 [INFO][68] monitor-addresses/startup.go 756: Using autodetected IPv4 address on interface eno1: 10.238.1.4/24
2020-10-28 09:30:09.897 [INFO][68] monitor-addresses/startup.go 756: Using autodetected IPv4 address on interface eno1: 10.238.1.4/24
2020-10-28 09:31:09.990 [INFO][68] monitor-addresses/startup.go 756: Using autodetected IPv4 address on interface eno1: 10.238.1.4/24
2020-10-28 09:32:09.993 [INFO][68] monitor-addresses/startup.go 756: Using autodetected IPv4 address on interface eno1: 10.238.1.4/24
2020-10-28 09:33:10.002 [INFO][68] monitor-addresses/startup.go 756: Using autodetected IPv4 address on interface eno1: 10.238.1.4/24
```

On some nodes pods keep running with a low CPU usage around 60m:

```
NAME                CPU(cores)   MEMORY(bytes)
calico-node-2x6b5   992m   171Mi
calico-node-4bwdr   61m   225Mi
calico-node-6h8bs   987m   124Mi
calico-node-d78r6   992m   172Mi
calico-node-ll6vs   990m   117Mi
calico-node-m9kxc   994m   172Mi
calico-node-p78n5   993m   116Mi
calico-node-pfz9d   995m   174Mi
calico-node-phfmn   60m   148Mi
calico-node-wc9ms   52m   235Mi
calico-node-xgzvb   998m   171Mi
calico-node-xwxl9   60m   153Mi
```

Usually, restarting the pod is sufficient to recover. 
This might be related to using route reflectors, since it popped up around disabling the full mesh and using 3 master nodes as dedicated route reflectors. 

## Steps to Reproduce (for bugs)

We have no simple steps to follow, but what we usually do on a weekly basis which triggers the bug: 
1. Redeploy 3 master nodes one-by-one, carrying calico-node, calico-typha and acting as route reflectors to the calico-mesh
2. Restart multiple worker nodes in parallel
3. Observe a varying amount of worker nodes being throttled by K8s to 1 CPU, when the others are sitting around 60m CPU

We did not check so far how much CPU would be consumed if we wouldn't limit at 1 CPU. 

## Your Environment

* Calico version `3.16.3`
* Orchestrator version (e.g. kubernetes, mesos, rkt): Kubernetes `1.18.10`
* Operating System and version: `FlatCar Container Linux 2605.6.0`

**Comment by mgleung** [comment]
Hi @sfudeus , thanks for reporting this. Just to clarify, are all the high CPU calico-nodes running on restarted worker nodes or does this  happen on some of the master nodes that are configured as route reflectors too? Also, could you try setting the Felix logs to `Debug` and see if that reveals anything? You can set it by setting `logSeverityScreen` in the FelixConfiguration resource (https://docs.projectcalico.org/reference/resources/felixconfig).

**Comment by sfudeus** [comment]
@mgleung As far as I have seen, it does not happen on the route reflectors. Will check the debug level. I am still trying to reproduce the issue for a single node - without redeploying the whole cluster.

**Comment by fasaxc** [comment]
Which process is using the CPU?  Is it the "Felix" process?  If so, please send it a SIGUSR2 and collect the CPU profile from `/tmp/` That should tell us what is using the CPU.

**Comment by sfudeus** [comment]
@fasaxc 
Yes, it is felix itself. 
Feel free to reach out for more info.

[felix-cpu-2020-11-05-10:06:48.pprof.gz](https://github.com/projectcalico/calico/files/5494247/felix-cpu-2020-11-05-10.06.48.pprof.gz)

**Comment by fasaxc** [comment]
Looks like Felix is spending all its time in GC.  Do you have a memory limit on the calico/node pod?  If so, looks like it's too small.

**Comment by sfudeus** [comment]
There is a memory limit set to 400M, but none of the pods received an OOM event or restarted. But I'll gladly try that. The pods which behave fine are running nicely on <200m CPU and 250M-300M memory. 
Here an excerpt of a `kubectl top pod` from today, showing CPU and memory usage:
```
calico-node-2f6mg   994m         249Mi
calico-node-4qdpn   996m         250Mi
calico-node-5jcs7   995m         261Mi
calico-node-5t8k4   998m         268Mi
calico-node-68mqf   995m         244Mi
calico-node-6qljw   998m         237Mi
calico-node-6t4bz   996m         236Mi
calico-node-7j7q2   995m         265Mi
calico-node-7x8fb   995m         195Mi
calico-node-8hjcq   997m         255Mi
calico-node-94dmc   1002m        267Mi
calico-node-9f78c   997m         241Mi
calico-node-b7q4d   998m         262Mi
calico-node-bbf26   998m         261Mi
calico-node-blccg   998m         301Mi
```

Here an excerpt of the healthy ones: 
```
calico-node-qgnxg   174m         239Mi
calico-node-r72ww   132m         240Mi
calico-node-sc856   171m         255Mi
calico-node-sdktr   122m         240Mi
calico-node-sgg6m   137m         255Mi
calico-node-t684z   116m         268Mi
calico-node-t7wkk   183m         275Mi
calico-node-tsbjq   131m         257Mi
calico-node-ttdm8   129m         232Mi
calico-node-v2ctz   139m         246Mi
calico-node-vfb7m   134m         258Mi
calico-node-vg6ng   119m         239Mi
calico-node-vksxh   135m         269Mi
calico-node-vrhkb   187m         265Mi
calico-node-w9r7q   195m         260Mi
calico-node-wfh9b   125m         247Mi
calico-node-x5jpm   137m         243Mi
calico-node-x9lbl   158m         242Mi
calico-node-z9kk8   121m         241Mi
```

**Comment by sfudeus** [comment]
Raising CPU limit to 2 and memory to 1Gi, not really improving:
```
calico-node-29fdq   1992m        216Mi
calico-node-2s6sk   1997m        219Mi
calico-node-2shr6   1997m        185Mi
calico-node-8dqxs   1996m        217Mi
calico-node-8x4nk   1997m        224Mi
```

What we see here as well: 
When updating the DaemonSet (including the calico-node pods on the masters (i.e. with route reflector), many pods come up again with high CPU load. When deleting/evicting single pods manually, they usually come up with regular load.

**Comment by fasaxc** [comment]
Maybe grab a few more profiles to see if we can spot a pattern?  Separately, turning on debug logging should give a hint as to what felix is churning.

Is it possible that these nodes just have some pods or policy churn going on?  It's normal for Felix to use more CPU when it needs to make dataplane changes.

**Comment by sfudeus** [comment]
Two more profile from different nodes (with 2CPU and 1G RAM):
[felix-cpu-2020-11-05-15:51:32.pprof.gz](https://github.com/projectcalico/calico/files/5495428/felix-cpu-2020-11-05-15.51.32.pprof.gz)
[felix-cpu-2020-11-05-15:20:18.pprof.gz](https://github.com/projectcalico/calico/files/5495429/felix-cpu-2020-11-05-15.20.18.pprof.gz)

The nodes are running productive workloads, so there is some churn regarding pods, but nothing that volatile. 

Debug logs unfortunately produce very much output, even under normal load, I'll try to grab some.

**Comment by sfudeus** [comment]
I'm attaching a 30 second debug log with 78k lines from one node - unfortunately that did not have the high cpu load (only around 500m), but maybe you see a general thing going wrong in our setup. 

[calico-node-log.txt.gz](https://github.com/projectcalico/calico/files/5495983/calico-node-log.txt.gz)

Additionally I attach an info-log from the first 30 minutes of a calico-node, which was running with 2 CPU.
[calico-node-zmj5j-log-info.txt.gz](https://github.com/projectcalico/calico/files/5496157/calico-node-zmj5j-log-info.txt.gz)

**Comment by fasaxc** [comment]
Think I say _a_ bug in your log; Typha is re-sending Node resources that haven't changed.  These are then treated as no-ops by felix but it wastes resources and causes GC (and it's exactly what Typha is intended to prevent). I think I can see why this is happening; the Node resource has been updated to contain its resource version so duplicates end up having a trivial difference.

That issue would be equal across all nodes, though.

I can't see anything else obvious in the debug log.  It looks like you're making good use of network policy.  That can result in uneven load (for example if one node has a lot of pods with complex policy with selectors that match large numbers of other pods, of, if one node has high pod churn where others are static).

**Comment by fasaxc** [comment]
Fix committed and backported to v3.17 and v3.16:
* https://github.com/projectcalico/typha/pull/454
* https://github.com/projectcalico/typha/pull/455

**Comment by vuxuanlai** [comment]
Hi @fasaxc ,
I'm facing same issue on v3.17.2, as you can see the following:
![calico](https://user-images.githubusercontent.com/1458400/123361495-5bb17080-d599-11eb-8aca-f76f5a3b634b.jpg)
Could you give me some reason?

**Comment by fasaxc** [comment]
@vuxuanlai What version of the typha pod are you running?

**Comment by JornShen** [comment]
we face the same problems. And watch the pporf,  also spend much cpu to execute gc. 
we use calico 3.18.1 this version. 
@fasaxc is the above MR no effect?

**Comment by fasaxc** [comment]
Please try the current release: v3.21.  does that solve your problem?  If not, please open a new issue; the issue tracked here has been fixed.

**Comment by naah69** [comment]
I'm facing same issue on v3.22.0

**Comment by fasaxc** [comment]
@naah69 Please try the latest Calico version (v3.24.x) at time of writing.  If the problem still occurs please open a new issue.  The one reported here was fixed.  Also, make sure you upgrade all components (including typha).
